### PR TITLE
GDrive Connector with Compass

### PR DESCRIPTION
**Pull request by giannis2two**
Adding Google Drive connector with Compass support

**AI Description**

This PR makes changes to the backend of the application. 

## Summary
- Adds a new ``Compass` tool to interact with a Compass instance.
- Updates the ``Google Drive` tool to include Compass logic.
- Introduces a ``CompassParserClient` and ``CompassClient` to interact with the CompassParser and Compass APIs respectively.
- Updates the ``BaseChat` class to include a return type annotation.
- Updates the ``update_variable_prompt` function to remove redundant parentheses.
- Updates the ``BaseTool` and ``BaseAuth` classes to include return type annotations for the ``is_available` and ``get_auth_url` methods.
- Updates the ``ToolAuth` model to include an `expires_at` field.
- Updates the ``ToolName` model to set the `is_visible` field to `True.
- Updates the ``black` linter to target Python 3.11.
- Updates the ``ruff-pre-commit` linter to version 0.4.10.
- Updates various package versions in ``poetry.lock`.

## Details
- **`.pre-commit-config.yaml`**: Updates the ``black` linter to target Python 3.11 and the ``ruff-pre-commit` linter to version 0.4.10.
- **`docker-compose.yml`**: Removes the ``POSTGRES_DB` environment variable from the ``toolkit` and ``test_toolkit` services.
- **`poetry.lock`**: Updates various package versions.
- **`src/backend/chat/base.py`**: Updates the ``BaseChat` class to include a return type annotation for the ``chat` method.
- **`src/backend/cli/main.py`**: Updates the ``update_variable_prompt` function to remove redundant parentheses.
- **`src/backend/compass_sdk/__init__.py`**: Adds a new ``Compass` tool to interact with a Compass instance.
- **`src/backend/compass_sdk/compass.py`**: Adds a new ``CompassClient` class to interact with the Compass API.
- **`src/backend/compass_sdk/constants.py`**: Adds various constants for the Compass SDK.
- **`src/backend/compass_sdk/parser.py`**: Adds a new ``CompassParserClient` class to interact with the CompassParser API.
- **`src/backend/compass_sdk/utils.py`**: Adds utility functions for the Compass SDK.
- **`src/backend/config/tools.py`**: Updates the ``ToolName` model to set the ``is_visible` field to `True.
- **`src/backend/crud/tool_auth.py`**: Updates the ``ToolAuth` model to include an `expires_at` field.
- **`src/backend/model_deployments/base.py`**: Updates the ``BaseDeployment` class to include return type annotations for the ``rerank_enabled`, ``list_models`, and ``is_available` methods.
- **`src/backend/schemas/chat.py`**: Removes the ``user_id` field from the ``BaseChatRequest` model.
- **`src/backend/schemas/tool_auth.py`**: Adds a new ``UpdateToolAuth` model for updating tool authentication details.
- **`src/backend/services/chat.py`**: Updates the ``process_chat` and ``generate_chat_stream` functions to improve error handling and formatting.
- **`src/backend/services/compass.py`**: Updates the ``Compass` tool to include Compass logic.
- **`src/backend/services/metrics.py`**: Updates the ``MetricsMiddleware` class to remove redundant parentheses.
- **`src/backend/tests/services/test_compass.py`**: Adds a test for the ``Compass` tool.
- **`src/backend/tools/base.py`**: Updates the ``BaseTool` and ``BaseAuth` classes to include return type annotations for the ``is_available` and ``get_auth_url` methods.
- **`src/backend/tools/compass.py`**: Removes the ``CompassTool` class.
- **`src/backend/tools/google_drive/__init__.py`**: Updates the ``Google Drive` tool imports.
- **`src/backend/tools/google_drive/auth.py`**: Updates the ``GoogleDriveAuth` class to use the new ``UpdateToolAuth` model.
- **`src/backend/tools/google_drive/constants.py`**: Adds various constants for the Google Drive tool.
- **`src/backend/tools/google_drive/tool.py`**: Updates the ``GoogleDrive` tool to include Compass logic.
- **`src/backend/tools/google_drive/utils.py`**: Adds utility functions for the Google Drive tool.
- **`src/backend/tools/utils/async_download.py`**: Adds a new module for asynchronously downloading files.
- **`src/backend/tools/utils/parallel_get_files.py`**: Adds a new module for parallel file retrieval.

### CODE REVIEW

File: src/backend/tools/google_drive/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from .auth import GoogleDriveAuth
```
Comment (giannis2two) [inline src/backend/tools/google_drive/__init__.py:1]: Moving `google_drive.py` logic under a `google_drive/` directory with the following file breakdowns

File: src/backend/tools/google_drive/auth.py [hunk 2]
```diff
@@ -74,7 +28,12 @@ def get_auth_url(cls, user_id: str) -> str:
         params = {
             "response_type": "code",
             "client_id": os.getenv("GOOGLE_DRIVE_CLIENT_ID"),
-            "scope": "https://www.googleapis.com/auth/drive",
+            "scope": " ".join(
+                [
+                    "https://www.googleapis.com/auth/drive",
+                    "https://www.googleapis.com/auth/drive.activity.readonly",
```
Comment (giannis2two) [inline src/backend/tools/google_drive/auth.py:34]: Adding drive activity scope for future syncing functionality

File: src/backend/tools/google_drive/auth.py [hunk 3]
```diff
@@ -74,7 +28,12 @@ def get_auth_url(cls, user_id: str) -> str:
         params = {
             "response_type": "code",
             "client_id": os.getenv("GOOGLE_DRIVE_CLIENT_ID"),
-            "scope": "https://www.googleapis.com/auth/drive",
+            "scope": " ".join(
+                [
+                    "https://www.googleapis.com/auth/drive.readonly",
```
Comment (giannis2two) [inline src/backend/tools/google_drive/auth.py:33]: More restrictive permission for just readonly access vs extra editting access

File: src/backend/schemas/tool_auth.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import datetime
+from typing import Optional
+
+from pydantic import BaseModel
+
+
+class UpdateToolAuth(BaseModel):
```
Comment (giannis2two) [inline src/backend/schemas/tool_auth.py:7]: Important to make GDrive `try_refresh_token` method not error

File: src/backend/tools/google_drive/auth.py [hunk 5]
```diff
@@ -118,13 +78,15 @@ def try_refresh_token(
             return False
         tool_auth_crud.update_tool_auth(
             session,
-            ToolAuth(
+            tool_auth,
+            UpdateToolAuth(
```
Comment (giannis2two) [inline src/backend/tools/google_drive/auth.py:82]: @BeatrixCohere follow up from your Google Drive PR, this addition is important for the `try_refresh_token` method. `update_tool_auth` expects a `BaseModel` class that inherits the `model_dump()` method https://github.com/cohere-ai/cohere-toolkit/blob/main/src/backend/crud/tool_auth.py#L58

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -10,11 +10,9 @@ repos:
     hooks:
     -   id: isort
         args: [--profile=black, --line-width=88, --force-grid-wrap=0, --use-parentheses, --float-to-top]
--   repo: https://github.com/psf/black
-    rev: 22.10.0
+-   repo: https://github.com/astral-sh/ruff-pre-commit
```
[the remark this answers is not in the document]
Comment (lusmoura) [reply .pre-commit-config.yaml:13]: You probably need to run `make lint` before pushing your changes!

**Comment by lusmoura** [comment]
To merge the PR, your commits need to be signed 😅 Here's how you can do it for old commits: https://superuser.com/questions/397149/can-you-gpg-sign-old-commits

You'll need to set it up: https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

And I recommend running `git config --global commit.gpgsign true` so that your next commits are signed

File: Makefile [hunk 7]
```diff
@@ -38,4 +38,6 @@ first-run:
 win-first-run:
 	make win-setup
 	make migrate
-	make dev
\ No newline at end of file
+	make dev
+backend-logs:
```
[the remark this answers is not in the document]
Comment (giannis2two) [reply Makefile:42]: I need to learn how to read..

[the remark this answers is not in the document]
Comment (giannis2two) [reply Makefile:42]: Removing backend-logs

File: src/backend/tools/google_drive/auth.py [hunk 8]
```diff
@@ -74,7 +29,12 @@ def get_auth_url(cls, user_id: str) -> str:
         params = {
             "response_type": "code",
             "client_id": os.getenv("GOOGLE_DRIVE_CLIENT_ID"),
-            "scope": "https://www.googleapis.com/auth/drive",
+            "scope": " ".join(
```
[the remark this answers is not in the document]
Comment (giannis2two) [reply src/backend/tools/google_drive/auth.py:32]: Yes thanks

**Comment by giannis2two** [comment]
> To merge the PR, your commits need to be signed 😅 Here's how you can do it for old commits: https://superuser.com/questions/397149/can-you-gpg-sign-old-commits
> 
> You'll need to set it up: https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification
> 
> And I recommend running `git config --global commit.gpgsign true` so that your next commits are signed

New to this, I will look inot it

File: .pre-commit-config.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (giannis2two) [reply .pre-commit-config.yaml]: As discussed in DMs, this is a weird issue Luisa hasn't seen before. `ruff` CLI suggested upgrading versions `v0.0.94 -> v0.4.10` and doing so does the trick

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -15,6 +15,6 @@ repos:
     hooks:
     -   id: black
 -   repo: https://github.com/charliermarsh/ruff-pre-commit
-    rev: v0.0.94
+    rev: v0.4.10
```
Comment (giannis2two) [inline .pre-commit-config.yaml:18]: See here https://github.com/cohere-ai/cohere-toolkit/pull/277#discussion_r1654911233

File: src/backend/chat/base.py [hunk 10]
```diff
@@ -8,6 +8,5 @@ class BaseChat:
     """Base for all chat options."""
 
     @abstractmethod
-    def chat(
-        self, chat_request: Union[CohereChatRequest, None], **kwargs: Any
-    ) -> Any: ...
+    def chat(self, chat_request: Union[CohereChatRequest, None], **kwargs: Any) -> Any:
```
Comment (giannis2two) [inline src/backend/chat/base.py:11]: make lint

File: src/backend/cli/main.py [hunk 11]
```diff
@@ -143,9 +143,7 @@ def update_variable_prompt(_, variables_to_update):
         variable_to_update = variable_to_update.split(":")[0]
         new_value = inquirer.text(f"Enter the new value for {variable_to_update}")
         write_env_file({variable_to_update: new_value})
-        print_styled(
-            f"🪛 Updated {variable_to_update} to {new_value}.", bcolors.OKGREEN
-        )
+        print_styled(f"🪛 Updated {variable_to_update} to {new_value}.", bcolors.OKGREEN)
```
Comment (giannis2two) [inline src/backend/cli/main.py:146]: make lint

File: src/backend/model_deployments/base.py [hunk 12]
```diff
@@ -19,23 +18,29 @@ class BaseDeployment:
 
     @property
     @abstractmethod
-    def rerank_enabled(self) -> bool: ...
+    def rerank_enabled(self) -> bool:
```
Comment (giannis2two) [inline src/backend/model_deployments/base.py:21]: make lint

File: src/backend/schemas/chat.py [hunk 13]
```diff
@@ -290,7 +290,6 @@ class ChatResponseEvent(BaseModel):
 
 
 class BaseChatRequest(BaseModel):
-
```
Comment (giannis2two) [inline src/backend/schemas/chat.py:293]: make lint

File: src/backend/routers/chat.py [hunk 14]
```diff
@@ -155,7 +155,6 @@ async def chat(
 def langchain_chat_stream(
     session: DBSessionDep, chat_request: LangchainChatRequest, request: Request
 ):
-
```
Comment (giannis2two) [inline src/backend/routers/chat.py:158]: make lint

File: src/backend/services/chat.py [hunk 15]
```diff
@@ -8,7 +8,6 @@
 from fastapi.encoders import jsonable_encoder
 from langchain_core.agents import AgentActionMessageLog
 from langchain_core.runnables.utils import AddableDict
-from starlette.exceptions import HTTPException
```
Comment (giannis2two) [inline src/backend/services/chat.py:11]: Unused

File: src/backend/services/chat.py [hunk 16]
```diff
@@ -88,7 +87,6 @@ def process_chat(
                 status_code=404, detail=f"Agent with ID {agent_id} not found."
             )
 
-        tool_names = [tool.name for tool in chat_request.tools]
```
Comment (giannis2two) [inline src/backend/services/chat.py:91]: Unused

File: src/backend/services/chat.py [hunk 17]
```diff
@@ -562,18 +560,21 @@ def generate_chat_stream(
...
-                conversation_id,
-                stream_end_data,
-                response_message,
-                document_ids_to_document,
-                session=session,
-                should_store=should_store,
-                user_id=user_id,
-                next_message_position=kwargs.get("next_message_position", 0),
-            )
+        (
```
Comment (giannis2two) [inline src/backend/services/chat.py:563]: make lint

File: src/backend/services/metrics.py [hunk 18]
```diff
@@ -269,9 +267,10 @@ def wrapper(self, chat_request: CohereChatRequest, **kwargs: Any) -> Any:
                     metrics_data.error = event_dict.get("error")
 
                 if event_dict.get("event_type") == StreamEvent.STREAM_END:
-                    metrics_data.input_nb_tokens, metrics_data.output_nb_tokens = (
-                        get_input_output_tokens(event_dict.get("response"))
-                    )
+                    (
```
Comment (giannis2two) [inline src/backend/services/metrics.py:270]: make lint

File: src/backend/tools/base.py [hunk 19]
```diff
@@ -13,10 +13,12 @@ class BaseTool:
 
     @classmethod
     @abstractmethod
-    def is_available(cls) -> bool: ...
+    def is_available(cls) -> bool:
```
Comment (giannis2two) [inline src/backend/tools/base.py:16]: make lint

File: .pre-commit-config.yaml [hunk 20]
```diff
@@ -14,6 +14,7 @@ repos:
     rev: 22.10.0
     hooks:
     -   id: black
+        args: ["--target-version", "py311"]
```
Comment (giannis2two) [inline .pre-commit-config.yaml:17]: Match our `pyproject.toml` python version

**Review by crclark-cohere — APPROVED** [review]

**Pull request merged**